orchestrator/careplancontributor/service.go

```diff
@@ -117,6 +145,112 @@ func (s Service) handleProxyToEPD(writer http.ResponseWriter, request *http.Requ
 	proxy.ServeHTTP(writer, request)
 }
 
+func (s Service) handleProxyToFHIR(writer http.ResponseWriter, request *http.Request) error {
+	// Authorize requester before proxying FHIR request
+	// Data holder must verify that the requester is part of the CareTeam by checking the URA
+	// Validate by retrieving the CarePlan from CPS, use URA in provided token to validate against CareTeam
+	// CarePlan should be provided in X-SCP-Context header
+	xScpContext := request.Header["X-Scp-Context"]
+	if len(xScpContext) == 0 {
+		// Check if header has different capitalisation, it may be changed by Go canonicalization
+		xScpContext = request.Header["X-SCP-Context"]
+		if len(xScpContext) == 0 {
+			err := errors.New("X-SCP-Context header value must be set")
+			http.Error(writer, err.Error(), http.StatusUnauthorized)
+			return err
+		}
+	}
+	if len(xScpContext) != 1 {
+		return errors.New("X-SCP-Context header must only contain one value")
+	}
+	carePlanURL := xScpContext[0]
+	if carePlanURL == "" {
+		err := errors.New("X-SCP-Context header value must be set")
+		http.Error(writer, err.Error(), http.StatusUnauthorized)
+		return err
+	}
+	u, err := url.Parse(carePlanURL)
+	if err != nil {
+		return err
+	}
+
+	// Get CarePlan from CPS
+	// TODO: How to go from link in X-SCP-Context to valid resource that can be passed to CPS?
+	carePlanServiceClient := fhirclient.New(s.carePlanServiceURL, s.scpHttpClient, coolfhir.Config())
+	var bundle fhir.Bundle
+	// Use extract CarePlan ID to be used for our query that will get the CarePlan and CareTeam in a bundle
+	carePlanId := strings.TrimPrefix(strings.TrimPrefix(u.Path, "/fhir/CarePlan/"), s.carePlanServiceURL.String())
+	newUrl := fmt.Sprintf("CarePlan\\?_id\\=%s\\&_include\\=CarePlan:care-team", carePlanId)
+	err = carePlanServiceClient.Read(newUrl, &bundle)
+	if err != nil {
+		http.Error(writer, err.Error(), http.StatusInternalServerError)
+		return err
+	}
+	// We are expecting a total of 1 result, with 2 entries (one for CarePlan, one for CareTeam)
+	if *bundle.Total == 0 {
+		err = errors.New("returned bundle has no results for CarePlan")
+		http.Error(writer, err.Error(), http.StatusNotFound)
+		return err
+	}
+	if len(bundle.Entry) != 2 {
+		err = errors.New(fmt.Sprintf("returned bundle has incorrect number of entries %d expecting 2", len(bundle.Entry)))
+		http.Error(writer, err.Error(), http.StatusInternalServerError)
+		return err
+	}
+
+	foundCareTeam := false
+	var careTeam fhir.CareTeam
+	// Get CareTeam from entries
+	for _, entry := range bundle.Entry {
+		err = json.Unmarshal(entry.Resource, &careTeam)
+		// Ignore errors, we are just trying to find the CareTeam
+		if err != nil || len(careTeam.Participant) == 0 {
+			continue
+		}
+		foundCareTeam = true
+		break
+	}
```

Comment:
You might be able to replace this with the syntactic sugar of coolfhir:

```
var bundle fhir.Bundle
	var careTeam []fhir.CareTeam
	err := coolfhir.ResourcesInBundle(&bundle, coolfhir.EntryIsOfType("CareTeam"), careTeam)
```

Reply:
That does look nicer. Will see if it will work!